File: Source/lua-common/LuaTuple.cpp
//
// Created by 冯鸿杰 on 17/1/19.
//

#include "LuaTuple.h"
#include "LuaValue.h"
#include "LuaNativeClass.hpp"
#include "LuaObjectEncoder.hpp"
#include "LuaObjectDecoder.hpp"
#include <typeinfo>

using namespace cn::vimfung::luascriptcore;

DECLARE_NATIVE_CLASS(LuaTuple);

cn::vimfung::luascriptcore::LuaTuple::LuaTuple()
    :LuaObject()
{

}

LuaTuple::LuaTuple (LuaObjectDecoder *decoder)
    :LuaObject(decoder)
{
    int size = decoder -> readInt32();    
    for (int i = 0; i < size; i++)
    {
        LuaValue *item = dynamic_cast<LuaValue *>(decoder -> readObject());
        _returnValues.push_back(item);
    }
}

cn::vimfung::luascriptcore::LuaTuple::~LuaTuple()
{
    for (LuaValueList::iterator it = _returnValues.begin(); it != _returnValues.end() ; ++it)
    {
        LuaValue *value = *it;
        value -> release();
    }
}

std::string LuaTuple::typeName()
{
    static std::string name = typeid(LuaTuple).name();
    return name;
}

void LuaTuple::serialization (LuaObjectEncoder *encoder)
{
    LuaObject::serialization(encoder);
    
    encoder -> writeInt32((int)_returnValues.size());
    for (LuaValueList::iterator it = _returnValues.begin(); it != _returnValues.end(); ++it)
    {
        LuaValue *value = *it;
        encoder -> writeObject(value);
    }
}

long cn::vimfung::luascriptcore::LuaTuple::count()
{
    return _returnValues.size();
}

void cn::vimfung::luascriptcore::LuaTuple::addReturnValue(LuaValue *value)
{
    _returnValues.push_back(value);
    value -> retain();
}

cn::vimfung::luascriptcore::LuaValue* cn::vimfung::luascriptcore::LuaTuple::getReturnValueByIndex(int index)
{
    return _returnValues[index];
}

void cn::vimfung::luascriptcore::LuaTuple::push(LuaContext *context)
{
    for (LuaValueList::iterator it = _returnValues.begin(); it != _returnValues.end() ; ++it)
    {
        LuaValue *value = *it;
        value -> push(context);
    }
}

File: Source/Unity3D/UnityProject/Assets/Plugins/LuaScriptCore/LuaTuple.cs
using System;
using System.Collections.Generic;

namespace cn.vimfung.luascriptcore
{
	/// <summary>
	/// 元组，仅用于返回多个值到Lua中使用，同时，Lua中的多个返回值会转换为该类型对象返回原生层。
	/// </summary>
	public class LuaTuple : LuaBaseObject
	{
		/// <summary>
		/// 初始化
		/// </summary>
		public LuaTuple ()
		{
			
		}

		/// <summary>
		/// 初始化
		/// </summary>
		/// <param name="decoder">对象解码器</param>
		public LuaTuple (LuaObjectDecoder decoder)
			: base(decoder)
		{
			int size = decoder.readInt32 ();
			if (size > 0) 
			{
				for (int i = 0; i < size; i++)
				{
					LuaValue item = decoder.readObject () as LuaValue;
					if (item != null) 
					{
						_returnValues.Add (item);
					}
				}
			}
		}

		/// <summary>
		/// 序列化对象
		/// </summary>
		/// <param name="encoder">对象编码器.</param>
		public override void serialization(LuaObjectEncoder encoder)
		{
			base.serialization (encoder);

			encoder.writeInt32 (_returnValues.Count);
			foreach (LuaValue value in _returnValues) 
			{
				encoder.writeObject (value);
			}
		}

		private List<LuaValue> _returnValues = new List<LuaValue> ();

		/// <summary>
		/// 获取返回值数量
		/// </summary>
		/// <value>返回值数量.</value>
		public int count
		{
			get
			{
				return _returnValues.Count;
			}
		}

		/// <summary>
		/// 添加返回值
		/// </summary>
		/// <param name="value">返回值.</param>
		public void addRetrunValue(object value)
		{
			_returnValues.Add (new LuaValue (value));
		}

		/// <summary>
		/// 获取返回值
		/// </summary>
		/// <returns>返回值</returns>
		/// <param name="index">位置索引.</param>
		public object getReturnValueByIndex(int index)
		{
			return _returnValues [index];
		}
	}
}

